### ISSUE TITLE
Enhance Discord Notifications

### ISSUE THREAD

**Issue by bakerboy448** [Sonarr/Sonarr#3985]
Title....

Implement support for Discordnotifer.com which provides enhanced discord notifications.
Feature exists in radarr https://github.com/Radarr/Radarr/pull/4611

PR #3819  would implement this.

**Comment by markus101** [comment]
We're aware of the PR, don't need a GHI after the fact.

**Comment by bakerboy448** [comment]
But the feature does not exist in Sonarr, correct?
PR's are not necessarily feature requests, they are code changes, correct?
Thus would a feature request not be appropriate to add a feature?

Just some thoughts, not expecting nor do I need a reply.  

I'm operating off the assumption that the PR is not merged because it is not a feature that you believe you or Taloth will use or enjoy and thus it is not a feature that will be added for the community to use. I'd certainly hope that is not the case, but 🤷

**Comment by markus101** [comment]
Correct it does not.
They are not feature requests, but a feature request doesn't need to exist for a PR to be merged to Sonarr.
If the GHI added value, sure, but adding a GHI to say you're interested in an existing PR doesn't really add value, it's just noise.

If we saw no value in the change it'd be closed, at this point we just haven't reviewed it.

**Comment by markus101** [comment]
Updated the title and re-opening because this isn't as cut and dry as landing that PR. Like Radarr, we can implement most if not all of the functionality that Discord Notifier offers, without using a third party site same as Radarr has done here: https://github.com/Radarr/Radarr/pull/5043

I'd much rather have native options for this, because it removes confusion for users and avoids reliance on a third party service. I'm not immediately closing that PR, but unless there is a compelling reason to go that route over doing it directly in Sonarr we'll do that.

**Comment by bakerboy448** [comment]
While I don't disagree that purely native is ideal.  I'm not sure the features in said third party service can fully be implemented natively in Sonarr given Discord's current accessibility...or lack thereof....many of the prime features require a discord bot at this time. I'm assuming neither Team Radarr or Sonarr want to go down the route of creating and maintaining a discord bot, would that be accurate? (inherently that would then rely on a third party service, a servarr and/or skyhook bot...)

While the end goal is great, it is likely a pretty significant effort to implement and obviously there are various other issues to address in the interim.   Would it not make sense to roll this into two phases?  Phase 1, implement the PR to provide immediate proven functionality.  Phase 2, work on implementing that same functionality natively. 

The PR is a proven concept that is soon going to flow to the remaining two downstream Arrs (Lidarr/Readarr)....and it would seem just plain silly for 1 of the 4 'siblings' not have a feature that is in the other 3 and extra silly when there is a PR ready for review at that. Would you not agree?

Also there's quite a few Connect options that rely on third parties...depending on how you frame the viewpoint technically most of Connect should not exist since it relies on third parties....

**Comment by markus101** [comment]
What features of the Discord Notifier notifications require a bot?
No, we're not going to maintain a bot.

What does Discord Notifier need that isn't already supported in Webhooks? Maybe the API key field? That's something that seems reasonable to add if necessary (though it could be part of the URL). Which I think is how it works currently, does it not?

No I don't think it should be added if it's going to immediately be done natively.

> Also there's quite a few Connect options that rely on third parties...depending on how you frame the viewpoint technically most of Connect should not exist since it relies on third parties....

They're all third party to Sonarr, but that's not the point, they're not third party to the actual notification service, with a confusingly similar name. I wouldn't be surprised if Discord Notifier as a name would be frowned upon by Discord because of the name.

**Comment by bakerboy448** [comment]
>What features of the Discord Notifier notifications require a bot?
- channel specifics (i.e. health notifications to a different channel)
- reactions
- editing messages (avoids 24 (1x/episode) * 3 (grab, download, import) notifications for a season pack)
- colors

> What does Discord Notifier need that isn't already supported in Webhooks? Maybe the API key field? That's something that seems reasonable to add if necessary (though it could be part of the URL). Which I think is how it works currently, does it not?

A bot; also see above

>No I don't think it should be added if it's going to immediately be done natively. It's more silly to support a bunch of different notifiers that do the same thing for the same service.

Again, it is technologically impossible at this time to support the features natively.

>if we support Discord Notifier directly, and someone else decides to create BetterDiscordNotifier and wants a custom implementation we'd have to consider adding that one too, because we've already march down that path.

Strawman argument...you can apply the same logic to any other third-party software connected to sonarr.
The reply would then be: no, this is what is implemented at this time.

>They're all third party to Sonarr, but that's not the point, they're not third party to the actual notification service, with a confusingly similar name. I wouldn't be surprised if Discord Notifier as a name would be frowned upon by Discord because of the name.

Sounds like a problem for Discord if they so wish to address it. That is neither Radarr's nor Sonarr's problem.

So to recap:
- the features offered by the 3rd party service, who is also very responsive and regularly implements community feedback, are not possible to natively implement in sonarr via discord webhooks and require a bot
- team sonarr and team radarr will not be creating a bot to implement said features
- enhancing native discord implementation would not provide the same feature set due to technological limitations
- the developer proposed native solution would not meet existing community expectations
- the PR has been successfully implemented into Radarr and will soon be in Readarr and then Lidarr

**Comment by austinwbest** [comment]
>  austinwbest deleted the austinwbest:phantom-Feature-Connect-DiscordNotifier branch 12 seconds ago

Resolved. You can close this issue.

**Comment by markus101** [comment]
Like I said when I re-opened the issue, this is about more than your particular GHI, but okay...

> A bot; also see above

That doesn't answer my question, what does the webhook connection in Sonarr pointed at Discord Notifier not provide that requires a custom solution? Like Newznab/Torznab indexers it'd make a lot more sense to have presets for webhooks than it would to have a custom solution for every possible consumer. That's kind of the point of Webhooks, you specify the payload and they consume it.

> Strawman argument...you can apply the same logic to any other third-party software connected to sonarr.
> The reply would then be: no, this is what is implemented at this time.

Sure it's easy to say no, but why even open the door?

As for a bot:

> channel specifics (i.e. health notifications to a different channel)

Different connections can achieve that.

> reactions
> editing messages (avoids 72 notifications... (1x/episode) * 3 (grab, download, import) * ~24 episodes notifications for a season pack)

Not sure what reactions would be used for, but editing messages is an interesting concept.

> colors

Of?

At this point given the PR was aggressively closed while we were discussing and the lack of clarity as to why it needs to be a custom solution instead of using the existing Webhooks (with some enhancements if data is missing), this discussion is moot, but I'm leaving this open because:

1. Enhancing the built-in Discord notification makes sense
2. If we can better support Discord Notifier, without a custom solution using a preset in Webhooks then I'm open to it.

**Comment by bakerboy448** [comment]
>That doesn't answer my question, what does the webhook connection in Sonarr pointed at Discord Notifier not provide that requires a custom solution? Like Newznab/Torznab indexers it'd make a lot more sense to have presets for webhooks than it would to have a custom solution for every possible consumer. That's kind of the point of Webhooks, you specify the payload and they consume it.

Already answered (removing channel specific, as you said easy workaround)
-reactions (see details)
![image](https://user-images.githubusercontent.com/55419169/94628737-62b78680-0286-11eb-907e-a82a3561ecdf.png)
![image](https://user-images.githubusercontent.com/55419169/94628794-9692ac00-0286-11eb-9cf5-6beb502edd95.png)
-editing messages (avoids 72 notifications... (1x/episode) * 3 (grab, download, import) * ~24 episodes notifications for a season pack)
-colors (of the notifications....[see the examples](https://discordnotifier.com/))

I'm not sure exactly what else you're looking for....

>Sure it's easy to say no, but why even open the door?

why do anything?

>At this point given the PR was aggressively closed while we were discussing and the lack of clarity as to why it needs to be a custom solution instead of using the existing Webhooks (with some enhancements if data is missing), this discussion is moot, but I'm leaving this open because[...]

I'm thinking it was "aggressively closed" (respectfully disagree it was aggressive) due to the fact the PR was sitting unaddressed and uncommented for awhile...combined with the comments here that strongly indicate it would likely **never** be implemented.

Evidence of the need for the custom use has already been provided...it doesn't appear the features noted above are available with webhooks

>-Enhancing the built-in Discord notification makes sense
-If we can better support Discord Notifier, without a custom solution using a preset in Webhooks then I'm open to it.

-Agreed
- doesn't appear possible.

**Comment by austinwbest** [comment]
Aggressive? It's sat there for months and this conversation is not going to lead anywhere. Taloth asked 2 months ago what would be needed to use a webhook instead, I replied with everything in the custom script and me to revise how the payload was received. Nothing.

Now an argument is made there is a potential native solution, 2 months later.

The solution was pre-made for months and got no where, this native solution will not be any different & it requires man power to get to even be an option that the other didn't.

There was nothing aggressive about it. There was an obvious answer and a pointless debate. No sense in dragging it on 👍

Just saving the back & forth.

**Comment by markus101** [comment]
You’re still missing the point, whatever bot discord notifier does is different from HOW it gets its provided the data. I understand how the boy is different from Sonarr would send the information and I’m not talking about that. You’re just repeating what you’ve already said.

Go to Connect in Sonarr add new and select webhook, that will send a PUT or a POST to the URL, just like lots of applications can do. It’s meant for other clients to be able to add support (much like discord themselves did to interpret calls meant for Slack without every application out there having to do that work. So I’m asking what data is that missing that Discord Notification requires (I know it was used on the past) adding that information takes care of the issue where other apps want to display information in their own custom way because they can consume the API.

> why do anything?

You’re right, we should just close all the GHIs and do nothing, because that makes sense. Like anything we do things that enhance Sonarr, but that doesn’t mean we implement every feature or implement it exactly as described, it works the same way for PRs.

A solution for notifications designed around Sonarr and the other *arrs should be able to receive web hooks to achieve this, doing that would allow all of the *arrs to be unchanged and the single app receiving them handles the information provided as necessary.

Given we hadn’t reviewed the PR (though admittedly should have by now) it wasn’t 100% clear what the solution there was, but having briefly looked and seeing a native solution that on the surface covered that behaviour, we’re not going to merge a PR only to be replaced with something else.

You’re right as-is that PR wasn’t going to land, but why webhooks sent to Discord Notifier from Sonarr is a valid discussion regardless of that, so closing the PR and telling us to close this issue because of that is aggressive.

Webhooks sent from Sonarr aren’t a solution to everything, but often that’s because those implementations have an existing established API.

**Comment by austinwbest** [comment]
So the bigger point of a connect was to not need to alter the native notifications. If something needed added or tweaked i could do it in the connect without having to even think about all the others who use the connect and all that you guys would have to do is merge (as little effort/work as possible was the goal, putting it on me instead). Group 1 uses webhooks/discord notifications and Group 2 uses the notifier as an example. 

I tend to add most things users ask for, so lets say they ask for new content that requires 1 more field, or 10 more, or 100 more.. Now you have to take into consideration what that does to Group 1 so the feature set of Group 2 is directly impacted by Group 1. If they want tag names? Well that is a call back because the names are not available directly in the notification, only the ids. So do we impact Group 1 for that too? genres? profiles? I mean the specifics isn't relevant to the point. 

You have to balance the other notifications accordingly and rightfully so, i dont disagree with that. With a native notifier connect none of that matters because the groups are not effected by each other.

As for another comment made, every connect option that accepts POST/GET/PUT or whatever technically doesn't need a native one. Every single one of them could have used the webhook implementation with small tweaks to their end to map the sonarr field to their field name before processing the request. Did they? no. It is just easier and more convenient for a user to hit the direct connect instead of having to set things up.

Have users setup multiple connects to simply point to different channels, because that is fluid for the user right.. Every point can have an argument as to why not to do it (as that is the end goal it has seemed since the start of this over 3 months ago) but that increases the steps for the end user.

In the end, there are reasons to allow it. Only you can determine if it goes into Sonarr and it appears that has been made. I wont ask for 1 or 100 fields be added to the existing notifications as i dont want to impact others based on a feature set that is not related to them, that is illogical. Same as i wont ask users to setup multiple connects and configure multiple connects.

**Comment by markus101** [comment]
> As for another comment made, every connect option that accepts POST/GET/PUT or whatever technically doesn't need a native one. Every single one of them could have used the webhook implementation with small tweaks to their end to map the sonarr field to their field name before processing the request. Did they? no. It is just easier and more convenient for a user to hit the direct connect instead of having to set things up.

The thing is, besides `Custom Script` and `Webhook` notifications every one of those connections is an already established product that has a standardized API that everyone uses to send messages, but Discord Notifier is build to take the raw information from Sonarr/Radarr/etc and translate that into a structure that Discord can understand, unfortunately, instead of consuming webhooks (and expanding those as necessary) you opted for taking custom scripts and converting that structure to the payload.

> I wont ask for 1 or 100 fields be added to the existing notifications as i dont want to impact others based on a feature set that is not related to them, that is illogical.

Adding an API field to provide authentication, instead of having it in the URL isn't illogical.

> Same as i wont ask users to setup multiple connects and configure multiple connects.

If you consumed the webhooks you'd do what you do now, they'd set up one connection and be done, your end would do the work it's doing now. If you're talking about the data the webhook would actually send, then I again disagree, adding more information to that payload is trivial and has little "cost" and only further enhances other's use of them. I imagine this is what's being used for Plex notifications, either through Tautulli or directly using Plex's webhooks.

The other problem comes down to the naming, Discord Notifier is confusingly similar to Discord, a user that doesn't know better is going to pick one, possibly the "Notifier" one because they want notifications, then they start trying to enter their Discord email/password and it's not working, now they've exposed those credentials, even though you're not doing it for nefarious purposes and they need help setting things up, so we need to explain a number of times what the difference is. Attempting to use Discord Notifier should be a conscious decision, not one they make because they don't know better, having it as a "Preset" for Webhooks like we do for Newznab indexer presets would allow things to be mostly pre-populated and the setup instructions on your end would be responsible for guiding the user through that process and letting them know they need to enter their API key.

**Comment by austinwbest** [comment]
So I am not opposed to getting it to work via webhook connect, even tried it months ago. I wrote the code to support and convert it on the fly on my end. Right off the bat downloadId was needed. Then another field, and to avoid asking for every field one by one I asked for the simple route.. Make the webhook payload the same as a custom script (data wise, not naming). Then I would go from there. The response:.... Crickets.. Nothing..

So while it is so easy so say everything is so trivial and doable, actually doing it is another battle all together. Yall are busy, hell we all are so I get that. But if things won't get added, it can't be used as reasoning.

I may have even said it earlier in this thread too.. Keep it simple and make the payloads match, then I have something to work with.

If adding new fields is something you are willing to do or at least merge a PR if one comes up related to new fields then sure it could work. But track record shows different on this specific topic. That is all I can go by bud.

**Comment by markus101** [comment]
I took a quick look and it looks like we're missing the deleted file paths (`Sonarr_DeletedRelativePaths` and `Sonarr_DeletedPaths`) and the source path for the episode file (`Sonarr_EpisodeFile_SourcePath`). If you've noticed something else missing let me know and I'll get it added. If there are other fields needed in the future there we can discuss / review PRs, etc

**Comment by austinwbest** [comment]
My only changes that i was about to do on the radarr connect was make sure the scene name & size was passed in grab/download/upgrade because as it stands now they are not in all iirc & add the previous full quality (WEB-DL 1080p) to upgrade notifications (dont think that is there but would have to look)

**Comment by markus101** [comment]
Scene name and size are sent with all, for grabs it's under `Release` and download/upgrade under `EpisodeFile`. Not too hard to treat the old files like the new file and include all the same properties.

**Comment by austinwbest** [comment]
Guess the WIKI is outdated? https://github.com/Sonarr/Sonarr/wiki/Webhook

I will look at the fields for scene/size again.

Thinking it would be useful to have what it was upgraded from in the upgrade notification (there would only be a few differing fields i guess?)

**Comment by austinwbest** [comment]
Yes, the deleted file info.

**Comment by austinwbest** [comment]
Health checks would for sure be needed...

**Comment by austinwbest** [comment]
> Scene name and size are sent with all, for grabs it's under `Release` and download/upgrade under `EpisodeFile`. Not too hard to treat the old files like the new file and include all the same properties.

`sceneName` is not in release for grab
```
	[release]
		[quality] => WEBDL-1080p
		[qualityVersion] => 1
		[releaseGroup] => raws
		[releaseTitle] => Boruto: Naruto Next Generations S01E168 1080p WEB-DL AAC 2.0 x264-Erai-raws
		[indexer] => BeyondHD
		[size] => 1449551488
	[downloadClient] => qBittorrent
	[downloadId] => A51E122B5D61F815A7EC2919BD78E7118142FB01
	[eventType] => Grab
```
`size` is not in episodeFile for download
```
	[episodeFile]
		[id] => 53025
		[relativePath] => Season 1\Boruto - Naruto Next Generations - S01E168 - Training Begins! WEBDL-1080p.mkv
		[path] => T:\Deluge\Complete\TV Shows\[Erai-raws] Boruto - Naruto Next Generations - 168 [1080p][Multiple Subtitle].mkv
		[quality] => WEBDL-1080p
		[qualityVersion] => 1
		[releaseGroup] => Erai-raws
		[sceneName] => [Erai-raws] Boruto - Naruto Next Generations - 168 [1080p][Multiple Subtitle]
	[isUpgrade] => 
	[downloadClient] => qBittorrent
	[downloadId] => A51E122B5D61F815A7EC2919BD78E7118142FB01
	[eventType] => Download
```

And `releaseGroup` also doesn't match between notifications

That is from 5:45AM this morning using `Version 3.0.3.938`

**Comment by markus101** [comment]
> `sceneName` is not in release for grab

Release Title == Scene Name.

> `size` is not in episodeFile for download

This is actually missing.

> And releaseGroup also doesn't match between notifications

Because your indexer is messing with the names, unless someone spends some time improving release group parsing then that's going to be a problem if your indexer gets cute with naming.
